compiler/src/dotty/tools/dotc/ast/TreeInfo.scala

```diff
@@ -166,8 +166,6 @@ trait TreeInfo[T >: Untyped <: Type] { self: Trees.Instance[T] =>
 
   /** can this type be a type pattern? */
   def mayBeTypePat(tree: Tree): Boolean = unsplice(tree) match {
-    case AndTypeTree(tpt1, tpt2) => mayBeTypePat(tpt1) || mayBeTypePat(tpt2)
-    case OrTypeTree(tpt1, tpt2) => mayBeTypePat(tpt1) || mayBeTypePat(tpt2)
     case RefinedTypeTree(tpt, refinements) => mayBeTypePat(tpt) || refinements.exists(_.isInstanceOf[Bind])
     case AppliedTypeTree(tpt, args) => mayBeTypePat(tpt) || args.exists(_.isInstanceOf[Bind])
```

Comment:
We probably need some recursion on the args to go into `AndTypeTree` and `OrTypeTree` trees.  For example `List[b] | Option[b]` would fail as the patterns are nested.

```scala
    case AppliedTypeTree(tpt, args) => mayBeTypePat(tpt) || args.exists(_.isInstanceOf[Bind]) || args.exists(mayBeTypePat)
```

Reply:
We should also have some tests for this.

Reply:
The method `maybeTypePat` git dropped altogether in the end.